lumen/ai/agents.py

```diff
@@ -0,0 +1,382 @@
+import io
+
+from typing import Literal, Type
+
+import pandas as pd
+import panel as pn
+import param
+import yaml
+
+from panel.chat import ChatInterface
+from panel.viewable import Viewer
+from pydantic import BaseModel, create_model
+
+from ..base import Component
+from ..dashboard import load_yaml
+from ..pipeline import Pipeline
+from ..sources import FileSource, InMemorySource
+from ..transforms.sql import SQLTransform, Transform
+from ..views import hvPlotUIView
+from .embeddings import Embeddings
+from .llm import Llm
+from .memory import memory
+from .models import String, Table
+from .translate import param_to_pydantic
+
+
+class Agent(Viewer):
+    """
+    An Agent in Panel is responsible for handling a specific type of
+    query. Each agent consists of an LLM and optionally a set of
+    embeddings.
+    """
+
+    debug = param.Boolean(default=True)
+
+    embeddings = param.ClassSelector(class_=Embeddings)
+
+    interface = param.ClassSelector(class_=ChatInterface)
+
+    llm = param.ClassSelector(class_=Llm)
+
+    system_prompt = param.String()
+
+    response_model = param.ClassSelector(class_=BaseModel, is_instance=False, default=String)
+
+    user = param.String(default='Assistant')
+
+    requires = param.List(default=[], readonly=True)
+
+    provides = param.List(default=[], readonly=True)
+
+    __abstract = True
+
+    def __init__(self, **params):
+        if 'interface' not in params:
+            params['interface'] = ChatInterface(callback=self._chat_invoke)
+        super().__init__(**params)
+
+    def _chat_invoke(self, contents: list | str, user: str, instance: ChatInterface):
+        return self.invoke(contents)
+
+    def __panel__(self):
+        return self.interface
+
+    def _system_prompt_with_context(self, messages: list | str) -> str:
+        system_prompt = self.system_prompt
+        if self.embeddings:
+            context = self.embeddings.query(messages)
+            system_prompt += f"{system_prompt}\n### CONTEXT: {context}".strip()
+        return system_prompt
+
+    def invoke(self, messages: list | str):
+        message = None
+        system_prompt = self._system_prompt_with_context(messages)
+        for chunk in self.llm.stream(messages, system=system_prompt, response_model=self.response_model):
+            if not chunk:
+                continue
+            message = self.interface.stream(
+                chunk, user=self.user, message=message, replace=True
+            )
+        return message
+
+
+class SourceAgent(Agent):
```

Comment:
Needs to be massively refactored, also maybe this shouldn't be an agent at all or it should probably ask the user what they want to do, upload or provide a URL.